modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/chm/TestConcurrentHashMapMvPartitionStorage.java

```diff
@@ -199,51 +218,103 @@ private static BinaryRow read(
         assert timestamp == null ^ txId == null;
 
         if (versionChain == null) {
-            return null;
+            return ReadResult.empty();
         }
 
         if (timestamp == null) {
             BinaryRow binaryRow = versionChain.row;
 
             if (filter != null && !filter.test(binaryRow)) {
-                return null;
+                return ReadResult.empty();
             }
 
             if (versionChain.txId != null && !versionChain.txId.equals(txId)) {
                 throw new TxIdMismatchException(txId, versionChain.txId);
             }
 
-            return binaryRow;
+            boolean isWriteIntent = versionChain.ts == null;
+
+            if (isWriteIntent) {
+                return ReadResult.createFromWriteIntent(
+                        binaryRow,
+                        versionChain.txId,
+                        versionChain.commitTableId,
+                        versionChain.next != null ? versionChain.next.ts : null,
+                        versionChain.commitPartitionId
+                );
+            }
+
+            return ReadResult.createFromCommitted(binaryRow);
         }
 
         VersionChain cur = versionChain;
 
-        if (cur.begin == null) {
+        if (cur.ts == null) {
+            if (cur.next == null) {
+                // We only have a write-intent.
+                BinaryRow binaryRow = cur.row;
+
+                if (filter != null && !filter.test(binaryRow)) {
+                    return ReadResult.empty();
+                }
+
+                return ReadResult.createFromWriteIntent(binaryRow, cur.txId, cur.commitTableId, null,
+                        cur.commitPartitionId);
+            }
+
             cur = cur.next;
         }
 
+        return walkVersionChain(versionChain, timestamp, filter, cur);
+    }
+
+    private static ReadResult walkVersionChain(VersionChain chainHead, HybridTimestamp timestamp, @Nullable Predicate<BinaryRow> filter,
+            VersionChain firstCommit) {
+        boolean hasWriteIntent = chainHead.ts == null;
+
+        if (hasWriteIntent && timestamp.compareTo(firstCommit.ts) > 0) {
+            // It's the latest commit in chain, query ts is greater than commit ts and there is a write-intent.
+            // So we just return write-intent.
+            BinaryRow binaryRow = chainHead.row;
+
+            if (filter != null && !filter.test(binaryRow)) {
+                return ReadResult.empty();
+            }
+
+            HybridTimestamp latestCommitTs = chainHead.next != null ? firstCommit.ts : null;
+
+            return ReadResult.createFromWriteIntent(binaryRow, chainHead.txId, chainHead.commitTableId, latestCommitTs,
+                    chainHead.commitPartitionId);
+        }
+
+        VersionChain cur = firstCommit;
+
         while (cur != null) {
-            if (timestamp.compareTo(cur.begin) >= 0) {
+            int compareResult = timestamp.compareTo(cur.ts);
```

Comment:
Might be picky, but I truly believe that inlining this expression would make code simipler. When I see a pattern `x.compareTo(y) >= 0`, I subconsciously replace it with `x >= y`. Splitting evaluation and comparison makes it harder.